src/sagas/task.saga.test.js

```diff
@@ -0,0 +1,382 @@
+import { call, put, select } from 'redux-saga/effects';
+import { cloneableGenerator } from 'redux-saga/utils';
+import * as childProcess from 'child_process';
+import * as path from 'path';
+import chalkRaw from 'chalk';
+
+import {
+  launchDevServer,
+  taskRun,
+  taskAbort,
+  displayTaskComplete,
+  taskComplete,
+  getBaseProjectEnvironment,
+  getDevServerCommand,
+} from './task.saga';
+import {
+  attachTaskMetadata,
+  receiveDataFromTaskExecution,
+  completeTask,
+  loadDependencyInfoFromDisk,
+} from '../actions';
+import killProcessId from '../services/kill-process-id.service';
+import { getProjectById } from '../reducers/projects.reducer';
+import { getPathForProjectId } from '../reducers/paths.reducer';
+import findAvailablePort from '../services/find-available-port.service';
+import { PACKAGE_MANAGER_CMD } from '../services/platform.service';
+import { ipcRenderer } from 'electron';
+
+const chalk = new chalkRaw.constructor({ level: 3 });
+
+// we want to watch and ensure that caught errors
+// are logged to the error log
+global.console = { error: jest.fn() };
+
+jest.mock('electron', () => ({
+  ipcRenderer: {
+    send: jest.fn(),
+  },
+  remote: {
+    app: { getAppPath: () => '' },
+  },
+}));
+
+jest.mock('uuid/v1', () => () => 'mocked-uuid-v1');
+
+// we don't actually need to run these commands through
+// the terminal, so an object with similar shape to that
+// returned by `child_process.spawn` should be sufficient.
+// Mock the relevant methods so they can be spied on.
+const mockSpawn = processId => ({
+  pid: processId,
+  write: jest.fn(),
+  on: jest.fn(),
+  stdin: {
+    write: jest.fn(),
+    on: jest.fn(),
+  },
+  stdout: {
+    write: jest.fn(),
+    on: jest.fn(),
+  },
+  stderr: {
+    write: jest.fn(),
+    on: jest.fn(),
+  },
+});
+
+describe('task saga', () => {
+  beforeEach(() => {
+    window.process = { env: {} };
```

Comment:
Right, the idea is that the saga uses `window.process.env`, and that's technically undefined in the tests?

This might be a bit paranoid, but I wonder if we should do this conditionally:

```
if (!window.process) {
  window.process = { env: {} };
}
```

This way, if we ever move test runners, and the runner provides a window.process, we aren't clobbering it unconditionally?

Also, I think this makes more sense as a `beforeAll`? Shouldn't need to re-run for every test.

Finally (sorry for all this!), would be great to add a comment pointing out why this is needed (looks like it's for `getBaseProjectEnvironment`?)

Reply:
All good points. Changed to `beforeAll` and made conditional.